cirq-google/cirq_google/ops/coupler_pulse_test.py

```diff
@@ -0,0 +1,151 @@
+            padding_time=cirq.Duration(nanos=4),
+        )
+    )
+    eq.add_equality_group(
+        coupler_pulse.CouplerPulse(
+            hold_time=cirq.Duration(nanos=10),
+            coupling_mhz=25.0,
+            rise_time=cirq.Duration(nanos=18),
+            padding_time=cirq.Duration(nanos=40),
+        )
+    )
+
+
+def test_coupler_pulse_validation():
+    with pytest.raises(ValueError, match='Full rise time'):
+        _ = coupler_pulse.CouplerPulse(
+            hold_time=cirq.Duration(nanos=20), coupling_mhz=25.0, rise_time=cirq.Duration(nanos=10)
+        )
+    with pytest.raises(ValueError, match='hold_time must be between'):
+        _ = coupler_pulse.CouplerPulse(
+            hold_time=cirq.Duration(nanos=110),
+            coupling_mhz=25.0,
+            rise_time=cirq.Duration(nanos=120),
+        )
+    with pytest.raises(ValueError, match='hold_time must be between'):
+        _ = coupler_pulse.CouplerPulse(
+            hold_time=cirq.Duration(nanos=-10), coupling_mhz=25.0, rise_time=cirq.Duration(nanos=20)
+        )
+    with pytest.raises(ValueError, match='padding_time must be between'):
+        _ = coupler_pulse.CouplerPulse(
+            hold_time=cirq.Duration(nanos=10),
+            coupling_mhz=25.0,
+            rise_time=cirq.Duration(nanos=20),
+            padding_time=cirq.Duration(nanos=200),
+        )
+    with pytest.raises(ValueError, match='padding_time must be between'):
+        _ = coupler_pulse.CouplerPulse(
+            hold_time=cirq.Duration(nanos=10),
+            coupling_mhz=25.0,
+            rise_time=cirq.Duration(nanos=20),
+            padding_time=cirq.Duration(nanos=-20),
+        )
+    with pytest.raises(ValueError, match='rise_time must be between'):
+        _ = coupler_pulse.CouplerPulse(
+            hold_time=cirq.Duration(nanos=10),
+            coupling_mhz=25.0,
+            rise_time=cirq.Duration(nanos=102),
+        )
+
+
+def test_coupler_pulse_str_repr():
+    gate = coupler_pulse.CouplerPulse(
+        hold_time=cirq.Duration(nanos=10), coupling_mhz=25.0, rise_time=cirq.Duration(nanos=18)
+    )
+    assert (
+        str(gate)
+        == 'CouplerPulse(hold_time=10 ns, coupling_mhz=25.0, '
+        + 'rise_time=18 ns, padding_time=2500.0 ps)'
+    )
+    assert (
+        repr(gate)
+        == 'cirq_google.ops.coupler_pulse.CouplerPulse('
+        + 'hold_time=cirq.Duration(nanos=10), '
+        + 'coupling_mhz=25.0, '
+        + 'rise_time=cirq.Duration(nanos=18), '
+        + 'padding_time=cirq.Duration(picos=2500.0))'
+    )
+
+
+def test_coupler_pulse_circuit_diagram():
+    a, b = cirq.LineQubit.range(2)
+    gate = coupler_pulse.CouplerPulse(
+        hold_time=cirq.Duration(nanos=10), coupling_mhz=25.0, rise_time=cirq.Duration(nanos=18)
+    )
+    circuit = cirq.Circuit(gate(a, b))
+    cirq.testing.assert_has_diagram(
+        circuit,
+        """
+0: ───/‾‾(10 ns@25.0MHz)‾‾\\───
+      │
+1: ───/‾‾(10 ns@25.0MHz)‾‾\\───
+""",
```

Comment:
here, too, consider using a raw string literal to avoid escapes.

Reply:
Done